File: 01307. Verbal Arithmetic Puzzle/solution.js
/**
 * Problem: 1307. Verbal Arithmetic Puzzle
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 3 ms (Beats 100%)
 */

/**
 * Solves cryptarithmetic puzzle where words sum to result
 *
 * @param {string[]} words - Array of words to sum
 * @param {string} result - Result word
 *
 * @returns {boolean} True if puzzle is solvable
 */
const isSolvable = (words, result) => {
  // Combine all words and result into a single array
  const allWords = [...words, result]

  // Get unique letters from all words and result
  const letters = [...new Set(allWords.join(''))]

  // If more than 10 unique letters, impossible (only digits 0-9)
  if (letters.length > 10) return false

  // Set to store letters that cannot be zero (first letters of each word)
  const nonZero = new Set()

  // Mark first letters of each word as non-zero
  for (const word of allWords) {
    if (word.length > 1) nonZero.add(word[0])
  }

  // Map to store coefficient for each letter (based on position value)
  const coefficient = new Map()

  // Initialize coefficient for each letter to 0
  for (const char of letters) coefficient.set(char, 0)

  // Calculate coefficients from words (positive contribution)
  for (const word of words) {
    // Start with multiplier 1 for units place
    let multiplier = 1

    // Process from rightmost digit to leftmost
    for (let i = word.length - 1; i >= 0; i--) {
      // Add current multiplier to letter's coefficient
      coefficient.set(word[i], coefficient.get(word[i]) + multiplier)

      // Move to next place value (tens, hundreds, etc.)
      multiplier *= 10
    }
  }

  // Subtract coefficients from result (negative contribution)
  {
    let multiplier = 1

    // Process from rightmost digit to leftmost
    for (let i = result.length - 1; i >= 0; i--) {
      // Subtract current multiplier from letter's coefficient
      coefficient.set(result[i], coefficient.get(result[i]) - multiplier)

      // Move to next place value
      multiplier *= 10
    }
  }

  // Sort letters by absolute coefficient for better pruning
  // Larger coefficients first to reduce branching
  letters.sort(
    (a, b) => Math.abs(coefficient.get(b)) - Math.abs(coefficient.get(a))
  )

  // Number of unique letters
  const letterCount = letters.length

  // Array of coefficients in same order as sorted letters
  const coeffArray = letters.map((char) => coefficient.get(char))

  // Precompute suffix sums for pruning bounds
  const suffixMax = new Array(letterCount + 1).fill(0),
    suffixMin = new Array(letterCount + 1).fill(0)

  // Calculate suffix bounds from right to left
  for (let i = letterCount - 1; i >= 0; i--) {
    // Current coefficient value
    const currentCoeff = coeffArray[i]

    if (currentCoeff > 0) {
      // Positive coefficient: max when digit is 9, min when digit is 0
      suffixMax[i] = suffixMax[i + 1] + currentCoeff * 9
      suffixMin[i] = suffixMin[i + 1]
    } else {
      // Negative coefficient: max when digit is 0, min when digit is 9
      suffixMax[i] = suffixMax[i + 1]
      suffixMin[i] = suffixMin[i + 1] + currentCoeff * 9
    }
  }

  // Track used digits (0-9) to ensure uniqueness
  const usedDigits = new Array(10).fill(false)

  // Recursive backtracking to assign digits to letters
  const backtrack = (index, currentSum) => {
    // Base case: all letters assigned, sum must be zero
    if (index === letterCount) return currentSum === 0

    // Pruning: impossible to reach zero with remaining letters
    if (currentSum + suffixMin[index] > 0 || currentSum + suffixMax[index] < 0)
      return false

    // Current letter to assign
    const currentChar = letters[index]

    // Coefficient for current letter
    const currentCoeff = coeffArray[index]

    // Starting digit (0 unless letter is leading and cannot be zero)
    const startDigit = nonZero.has(currentChar) ? 1 : 0

    // Try all possible digits for current letter
    for (let digit = startDigit; digit <= 9; digit++) {
      // Skip if digit already used
      if (usedDigits[digit]) continue

      // Mark digit as used
      usedDigits[digit] = true

      // Recurse with updated sum
      if (backtrack(index + 1, currentSum + currentCoeff * digit)) return true

      // Backtrack: unmark digit
      usedDigits[digit] = false
    }

    // No valid digit found for current letter
    return false
  }

  // Start backtracking from first letter with sum 0
  return backtrack(0, 0)
}

File: 01307. Verbal Arithmetic Puzzle/solution.cs
/**
 * Problem: 1307. Verbal Arithmetic Puzzle
 *
 * Difficulty: Hard
 *
 * Language: C#
 *
 * Performance: Runtime - 5 ms (Beats 100%)
 */

public class Solution
{
  // Coefficient for each letter based on positional value
  private long[] coefficient;

  // List of letter indices sorted by absolute coefficient
  private List<int> sortedIndices;

  // Set of letters that cannot be zero (first letters of words)
  private HashSet<char> nonZeroLetters;

  // Current digit assignment for each letter index
  private int[] digitAssignment;

  // Track which digits (0-9) are already used
  private bool[] usedDigits;

  // Map from index back to character
  private char[] indexToChar;

  // Map from character to index
  private Dictionary<char, int> charToIndex;

  public bool IsSolvable(string[] words, string result)
  {
    // Initialize mapping from characters to indices
    charToIndex = new Dictionary<char, int>();
    int currentIndex = 0;

    // Assign indices to all characters in words
    foreach (string word in words)
    {
      foreach (char c in word)
      {
        if (!charToIndex.ContainsKey(c))
          charToIndex[c] = currentIndex++;
      }
    }

    // Assign indices to characters in result
    foreach (char c in result)
    {
      if (!charToIndex.ContainsKey(c))
        charToIndex[c] = currentIndex++;
    }

    // Number of unique letters
    int letterCount = charToIndex.Count;

    // Initialize coefficient array
    coefficient = new long[letterCount];

    // Initialize mapping from index to character
    indexToChar = new char[letterCount];
    foreach (var kv in charToIndex)
      indexToChar[kv.Value] = kv.Key;

    // Calculate coefficients from words (positive contribution)
    foreach (string word in words)
    {
      long multiplier = 1;

      // Process from rightmost digit to leftmost
      for (int i = word.Length - 1; i >= 0; i--)
      {
        coefficient[charToIndex[word[i]]] += multiplier;
        multiplier *= 10;
      }
    }

    // Subtract coefficients from result (negative contribution)
    long resultMultiplier = 1;
    for (int i = result.Length - 1; i >= 0; i--)
    {
      coefficient[charToIndex[result[i]]] -= resultMultiplier;
      resultMultiplier *= 10;
    }

    // Identify letters that cannot be zero (first letters)
    nonZeroLetters = new HashSet<char>();
    foreach (string word in words)
    {
      if (word.Length > 1)
        nonZeroLetters.Add(word[0]);
    }
    if (result.Length > 1)
      nonZeroLetters.Add(result[0]);

    // Create list of indices sorted by absolute coefficient
    sortedIndices = new List<int>();
    for (int i = 0; i < letterCount; i++)
      sortedIndices.Add(i);

    // Sort indices by descending absolute coefficient for better pruning
    sortedIndices.Sort((a, b) => Math.Abs(coefficient[b]).CompareTo(Math.Abs(coefficient[a])));

    // Initialize digit assignment and usage tracking
    digitAssignment = new int[letterCount];
    usedDigits = new bool[10];

    // Start depth-first search
    return DFS(0, 0);
  }

  /// <summary>
  /// Depth-first search to assign digits to letters
  /// </summary>
  /// <param name="position">Current position in sorted indices</param>
  /// <param name="currentSum">Current sum of assigned digits</param>
  /// <returns>True if solution found</returns>
  private bool DFS(int position, long currentSum)
  {
    // Base case: all letters assigned, sum must be zero
    if (position == sortedIndices.Count)
      return currentSum == 0;

    // Calculate remaining sum bounds for pruning
    int currentIndex = sortedIndices[position];
    long maxRemaining = 0,
      minRemaining = 0;

    // Calculate maximum and minimum possible sum from remaining letters
    for (int j = position; j < sortedIndices.Count; j++)
    {
      int idx = sortedIndices[j];
      long coeff = coefficient[idx];

      if (coeff > 0)
      {
        // Positive coefficient: max with digit 9, min with digit 0
        maxRemaining += coeff * 9;
        minRemaining += 0;
      }
      else
      {
        // Negative coefficient: max with digit 0, min with digit 9
        maxRemaining += 0;
        minRemaining += coeff * 9;
      }
    }

    // Prune if impossible to reach zero with remaining letters
    if (currentSum + maxRemaining < 0 || currentSum + minRemaining > 0)
      return false;

    // Current letter to assign
    char currentLetter = indexToChar[currentIndex];

    // Check if this letter cannot be zero
    bool isNonZero = nonZeroLetters.Contains(currentLetter);

    // Starting digit (0 unless letter is non-zero)
    int startDigit = isNonZero ? 1 : 0;

    // Try all possible digits for current letter
    for (int digit = startDigit; digit <= 9; digit++)
    {
      // Skip if digit already used
      if (usedDigits[digit])
        continue;

      // Assign digit
      usedDigits[digit] = true;

      // Recurse with updated sum
      if (DFS(position + 1, currentSum + coefficient[currentIndex] * digit))
        return true;

      // Backtrack
      usedDigits[digit] = false;
    }

    return false;
  }
}
